packages/next/server/base-server.ts

```diff
@@ -407,14 +409,21 @@ export default abstract class Server<ServerOptions extends Options = Options> {
         parsedUrl.query = parseQs(parsedUrl.query)
       }
 
+      const protocol = (
+        (req as NodeNextRequest).originalRequest?.socket as TLSSocket
+      )?.encrypted
```

Comment:
We potentially might want to support [`x-forwarded-proto`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto) in a follow-up but this will probably require config to verify the upstream proxy.